opentelemetry-api/src/opentelemetry/correlationcontext/__init__.py

```diff
@@ -0,0 +1,128 @@
+        """
+        To access the value for an entry by a prior event, the Correlations API provides a
+        function which takes a context and a name as input, and returns a value. Returns
+        the Value associated with the given Name, or null if the given Name is not present.
+
+        Args:
+            context: The context in which to find the CorrelationContext
+            name: The name of the entry to retrieve
+        Returns:
+            The value of the entry matching the name
+        """
+        # pylint: disable=no-self-use
+        correlation_context = context.get_value(CORRELATION_CONTEXT_KEY)
+        if correlation_context and name in correlation_context:
+            return correlation_context[name]
+        return None
+
+    def set_correlation(
+        self, context: ContextAPI, name: str, value: typing.Any
+    ) -> ContextAPI:
+        """
+        To record the value for an entry, the Correlations API provides a function which takes a
+        context, a name, and a value as input. Returns an updated Context which contains the new value.
+
+        Args:
+            context: The context in which to find the CorrelationContext
+            name: The name of the entry to set
+            value: The value of the entry
+        Returns:
+            A new Context object with updated correlations
+        """
+        new_correlation_context = self._copy()
+        new_correlation_context[name] = value
+        return context.set_value(
+            CORRELATION_CONTEXT_KEY, new_correlation_context
+        )
+
+    def remove_correlation(self, context: ContextAPI, name: str) -> ContextAPI:
+        """
+        To delete an entry, the Correlations API provides a function which takes a context and a name as
+        input. Returns an updated Context which no longer contains the selected Name.
+
+        Args:
+            context: The context in which to remove the CorrelationContext
+            name: The name of the entry to remove
+        Returns:
+            A new Context object with the correlation removed
+        """
+        new_correlation_context = self._copy()
+        if name in new_correlation_context:
+            del new_correlation_context[name]
+        return context.set_value(
+            CORRELATION_CONTEXT_KEY, new_correlation_context
+        )
+
+    def clear_correlations(self, context: ContextAPI) -> ContextAPI:
+        """
+        To avoid sending any entries to an untrusted process, the Correlation API provides a function
+        to remove all Correlations from a context. Returns an updated Context with no correlations.
+
+        Args:
+            context: The context in which to clear the CorrelationContext
+        Returns:
+            A new Context object with no correlations
+        """
+        # pylint: disable=no-self-use
+        return context.set_value(CORRELATION_CONTEXT_KEY, {})
+
+    def _copy(self) -> typing.Dict[str, typing.Any]:
```

Comment:
Since this will be called quite often I think we should keep performance in mind and try to use the stdlib contextvars on Python versions where this is possible (is it possible to nest contexts, i.e store a context in a context slot?). If this is not possible, we might be able to reuse ideas from https://www.python.org/dev/peps/pep-0567/#implementation to achieve a constant time copy function. Alternatively, maybe this is the point where we should really use the contextvars backport (see #101). The cool thing about contextvars is that it has an immutable context API, just like the OTEP's design.